mono/mini/mini-posix.c

```diff
@@ -521,123 +505,255 @@ mono_runtime_cleanup_handlers (void)
 	free_saved_signal_handlers ();
 }
 
-#ifdef HAVE_LINUX_RTC_H
-#include <linux/rtc.h>
-#include <sys/ioctl.h>
-#include <fcntl.h>
-static int rtc_fd = -1;
+#ifdef HAVE_PROFILER_SIGNAL
 
-static int
-enable_rtc_timer (gboolean enable)
+static gint64
+clock_get_time_ns (MonoProfileSamplingMode mode)
 {
-	int flags;
-	flags = fcntl (rtc_fd, F_GETFL);
-	if (flags < 0) {
-		perror ("getflags");
-		return 0;
-	}
-	if (enable)
-		flags |= FASYNC;
-	else
-		flags &= ~FASYNC;
-	if (fcntl (rtc_fd, F_SETFL, flags) == -1) {
-		perror ("setflags");
-		return 0;
+	struct timespec ts;
+
+#ifdef PLATFORM_MACOSX
+	kern_return_t ret;
+	clock_id_t mach_clock = SYSTEM_CLOCK;
+	clock_serv_t clock_service;
+	mach_timespec_t mach_ts;
+
+	if ((ret = host_get_clock_service (mach_host_self (), mach_clock, &clock_service)) != KERN_SUCCESS)
+		g_error ("%s: host_get_clock_service () returned %d", __func__, ret);
+
+	if ((ret = clock_get_time (clock_service, &mach_ts)) != KERN_SUCCESS)
+		g_error ("%s: clock_get_time () returned %d", __func__, ret);
+
+	if ((ret = mach_port_deallocate (mach_task_self (), clock_service)) != KERN_SUCCESS)
+		g_error ("%s: mach_port_deallocate () returned %d", __func__, ret);
+
+	ts.tv_sec = mach_ts.tv_sec;
+	ts.tv_nsec = mach_ts.tv_nsec;
+#else
+	clockid_t posix_clock;
+
+	switch (mode) {
+	case MONO_PROFILER_STAT_MODE_PROCESS:
+#ifdef HAVE_CLOCK_NANOSLEEP
+		/*
+		 * If we don't have clock_nanosleep (), measuring the process time
+		 * makes very little sense as we can only use nanosleep () to sleep on
+		 * real time.
+		 */
+		posix_clock = CLOCK_PROCESS_CPUTIME_ID;
+		break;
+#endif
+	case MONO_PROFILER_STAT_MODE_REAL: posix_clock = CLOCK_MONOTONIC; break;
+	default: g_assert_not_reached (); break;
 	}
-	return 1;
-}
+
+	if (clock_gettime (posix_clock, &ts) == -1)
+		g_error ("%s: clock_gettime () returned -1, errno = %d", __func__, errno);
 #endif
 
-void
-mono_runtime_shutdown_stat_profiler (void)
+	return ((gint64) ts.tv_sec * 1000000000) + (gint64) ts.tv_nsec;
+}
+
+static void
+clock_sleep_ns_abs (MonoProfileSamplingMode mode, guint64 ns_abs)
 {
-#ifdef HAVE_LINUX_RTC_H
-	if (rtc_fd >= 0)
-		enable_rtc_timer (FALSE);
+	MONO_PREPARE_BLOCKING;
```

Comment:
Good point, those were leftovers from when I was trying to abstract this stuff in `mono-time.c`.